packages/node-core/src/indexer/connectionPool.service.ts

```diff
@@ -1,57 +1,153 @@
 // Copyright 2020-2022 OnFinality Limited authors & contributors
 // SPDX-License-Identifier: Apache-2.0
 
-import {Injectable, OnApplicationShutdown} from '@nestjs/common';
+import {OnApplicationShutdown, Injectable} from '@nestjs/common';
+import {Interval} from '@nestjs/schedule';
+import chalk from 'chalk';
 import {toNumber} from 'lodash';
+import {IApi} from '..';
 import {getLogger} from '../logger';
 
-const logger = getLogger('api');
+const logger = getLogger('connection-pool');
+
+const MAX_FAILURES = 5;
+const LOG_INTERVAL_MS = 60 * 1000; // Log every 60 seconds
+const RESPONSE_TIME_WEIGHT = 0.7;
+const FAILURE_WEIGHT = 0.3;
+const RETRY_DELAY = 60 * 1000;
 
 export interface ApiConnection {
-  apiConnect(): Promise<void>;
-  apiDisconnect(): Promise<void>;
+  //apiConnect(): Promise<void>;
+  //apiDisconnect(): Promise<void>;
+  api: any;
+}
+
+export enum ApiErrorType {
+  Timeout = 'timeout',
+  Connection = 'connection',
+  Default = 'default',
+}
+
+export class ApiConnectionError extends Error {
+  errorType: ApiErrorType;
+
+  constructor(name: string, message: string, errorType: ApiErrorType) {
+    super(message);
+    this.name = name;
+    this.errorType = errorType;
+  }
 }
 
 @Injectable()
-export class ConnectionPoolService<T extends ApiConnection> implements OnApplicationShutdown {
+export class ConnectionPoolService<T extends IApi<any, any, any>> implements OnApplicationShutdown {
   private allApi: T[] = [];
+  private apiToIndexMap: Map<any, number> = new Map();
+  private indexToEndpointMap: Record<number, string> = {};
   private connectionPool: Record<number, T> = {};
-  private taskCounter = 0;
+  private performanceScores: Record<number, number> = {};
+  private backoffDelays: Record<number, number> = {};
+  private failureCounts: Record<number, number> = {};
```

Comment:
Rather than having a bunch of mappings it might make sense to have a class/object that contains all of these into one.